### ISSUE TITLE
Plans to Support Apple Silicon builds?

### ISSUE THREAD

**Issue by eengoron** [electron-userland/electron-builder#5392]
Are there current plans to support Apple Silicon builds with `electron-builder`? We're currently in the process of making sure our electron app is ready for ARM support, so I wanted to make sure that we'd be able to build to those targets using `electron-builder`.

If so, what does the release timeline look like for that? 

Since https://github.com/electron-userland/electron-builder/issues/5095 was closed with no resolution, I figured it best to open a new issue.

**Comment by ahmadwaliesipick** [comment]
any update on this?

**Comment by ahmadwaliesipick** [comment]
> Are there current plans to support Apple Silicon builds with `electron-builder`? We're currently in the process of making sure our electron app is ready for ARM support, so I wanted to make sure that we'd be able to build to those targets using `electron-builder`.
> 
> If so, what does the release timeline look like for that?
> 
> Since #5095 was closed with no resolution, I figured it best to open a new issue.

@eengoron  see here they have put this ticket in backlog and marked closed even still electron-builder is not supporting ARM64 build. #5095

**Comment by eengoron** [comment]
@ahmadwaliesipick -- I mentioned that ticket in my original message. That ticket was closed without resolution, so I opened a new one.

cc @develar

**Comment by ahmadwaliesipick** [comment]
> @ahmadwaliesipick -- I mentioned that ticket in my original message. That ticket was closed without resolution, so I opened a new one.
> 
> cc @develar

Yes @eengoron . Do you have any information are they working on ARM64 build?

**Comment by aabuhijleh** [comment]
Unfortunately, it looks like it's time to consider migrating to electron-packager for this issue :(

**Comment by develar** [comment]
Anyone is willing to open a pull request? Fix should be easy, as far I see.

**Comment by deathlyrage** [comment]
Can we get this issued pinned, saves all the duplicate issues being made and easier to focus for people helping out with pull requests and debugging.

**Comment by mmaietta** [comment]
FWIW, I was able to build an `arm64` electron example app with the help of `patch-package` on two files.  Perhaps this will help/unblock people investigating further 🙂 

I was able to package via both `yarn electron-builder` to use package.json and cli `yarn electron-builder --dir --arm64`

Unfortunately, I'm actually unable to test the app though as I don't have a compatible MacOS version installed right now. Just to be verbose, here's the error alert
> You can’t use this version of the application “MyApp.app” with this version of macOS.
You have macOS 10.15.7. The application requires macOS 11.0 or later.

Maybe someone would be willing to test this? 😉 

---

Here are the changes I made:

package.json excerpt:
```
"build": {
    "appId": "test",
    "mac": {
      "target" : { "target": "dir", "arch": "arm64" }
    }
  },
  "devDependencies": {
    "electron": "11.0.0-beta.20",
    "electron-builder": "^22.9.1"
  },
```
Patch 1:
`app-builder-lib+22.9.1.patch`
```
diff --git a/node_modules/app-builder-lib/out/macPackager.js b/node_modules/app-builder-lib/out/macPackager.js
index 4825f78..2aa682b 100644
--- a/node_modules/app-builder-lib/out/macPackager.js
+++ b/node_modules/app-builder-lib/out/macPackager.js
@@ -258,7 +258,7 @@ class MacPackager extends _platformPackager().PlatformPackager {
 
     if (!hasMas || targets.length > 1) {
       const appPath = prepackaged == null ? path.join(this.computeAppOutDir(outDir, arch), `${this.appInfo.productFilename}.app`) : prepackaged;
-      nonMasPromise = (prepackaged ? Promise.resolve() : this.doPack(outDir, path.dirname(appPath), this.platform.nodeName, arch, this.platformSpecificBuildOptions, targets)).then(() => this.packageInDistributableFormat(appPath, _builderUtil().Arch.x64, targets, taskManager));
+      nonMasPromise = (prepackaged ? Promise.resolve() : this.doPack(outDir, path.dirname(appPath), this.platform.nodeName, arch, this.platformSpecificBuildOptions, targets)).then(() => this.packageInDistributableFormat(appPath, arch, targets, taskManager));
     }
 
     for (const target of targets) {

```
Patch 2:
`electron-builder+22.9.1.patch`
```
diff --git a/node_modules/electron-builder/out/builder.js b/node_modules/electron-builder/out/builder.js
index 59383a1..8b9aef7 100644
--- a/node_modules/electron-builder/out/builder.js
+++ b/node_modules/electron-builder/out/builder.js
@@ -69,10 +69,6 @@ function normalizeOptions(args) {
 
   function processTargets(platform, types) {
     function commonArch(currentIfNotSpecified) {
-      if (platform === _appBuilderLib().Platform.MAC) {
-        return args.x64 || currentIfNotSpecified ? [_builderUtil().Arch.x64] : [];
-      }
-
       const result = Array();
 
       if (args.x64) {
@@ -242,7 +238,7 @@ function createTargets(platforms, type, arch) {
   const targets = new Map();
 
   for (const platform of platforms) {
-    const archs = platform === _appBuilderLib().Platform.MAC ? [_builderUtil().Arch.x64] : arch === "all" ? [_builderUtil().Arch.x64, _builderUtil().Arch.ia32] : [(0, _builderUtil().archFromString)(arch == null ? process.arch : arch)];
+    const archs = arch === "all" && platform !== _appBuilderLib().Platform.MAC ? [_builderUtil().Arch.x64, _builderUtil().Arch.ia32] : [(0, _builderUtil().archFromString)(arch == null ? process.arch : arch)];
     const archToType = new Map();
     targets.set(platform, archToType);
 

```

**Comment by ahmadwaliesipick** [comment]
> FWIW, I was able to build an `arm64` electron example app with the help of `patch-package` on two files. Perhaps this will help/unblock people investigating further 🙂
> 
> I was able to package via both `yarn electron-builder` to use package.json and cli `yarn electron-builder --dir --arm64`
> 
> Unfortunately, I'm actually unable to test the app though as I don't have a compatible MacOS version installed right now. Just to be verbose, here's the error alert
> 
> > You can’t use this version of the application “MyApp.app” with this version of macOS.
> > You have macOS 10.15.7. The application requires macOS 11.0 or later.
> 
> Maybe someone would be willing to test this? 😉
> 
> Here are the changes I made:
> 
> package.json excerpt:
> 
> ```
> "build": {
>     "appId": "test",
>     "mac": {
>       "target" : { "target": "dir", "arch": "arm64" }
>     }
>   },
>   "devDependencies": {
>     "electron": "11.0.0-beta.20",
>     "electron-builder": "^22.9.1"
>   },
> ```
> 
> Patch 1:
> `app-builder-lib+22.9.1.patch`
> 
> ```
> diff --git a/node_modules/app-builder-lib/out/macPackager.js b/node_modules/app-builder-lib/out/macPackager.js
> index 4825f78..2aa682b 100644
> --- a/node_modules/app-builder-lib/out/macPackager.js
> +++ b/node_modules/app-builder-lib/out/macPackager.js
> @@ -258,7 +258,7 @@ class MacPackager extends _platformPackager().PlatformPackager {
>  
>      if (!hasMas || targets.length > 1) {
>        const appPath = prepackaged == null ? path.join(this.computeAppOutDir(outDir, arch), `${this.appInfo.productFilename}.app`) : prepackaged;
> -      nonMasPromise = (prepackaged ? Promise.resolve() : this.doPack(outDir, path.dirname(appPath), this.platform.nodeName, arch, this.platformSpecificBuildOptions, targets)).then(() => this.packageInDistributableFormat(appPath, _builderUtil().Arch.x64, targets, taskManager));
> +      nonMasPromise = (prepackaged ? Promise.resolve() : this.doPack(outDir, path.dirname(appPath), this.platform.nodeName, arch, this.platformSpecificBuildOptions, targets)).then(() => this.packageInDistributableFormat(appPath, arch, targets, taskManager));
>      }
>  
>      for (const target of targets) {
> ```
> 
> Patch 2:
> `electron-builder+22.9.1.patch`
> 
> ```
> diff --git a/node_modules/electron-builder/out/builder.js b/node_modules/electron-builder/out/builder.js
> index 59383a1..8b9aef7 100644
> --- a/node_modules/electron-builder/out/builder.js
> +++ b/node_modules/electron-builder/out/builder.js
> @@ -69,10 +69,6 @@ function normalizeOptions(args) {
>  
>    function processTargets(platform, types) {
>      function commonArch(currentIfNotSpecified) {
> -      if (platform === _appBuilderLib().Platform.MAC) {
> -        return args.x64 || currentIfNotSpecified ? [_builderUtil().Arch.x64] : [];
> -      }
> -
>        const result = Array();
>  
>        if (args.x64) {
> @@ -242,7 +238,7 @@ function createTargets(platforms, type, arch) {
>    const targets = new Map();
>  
>    for (const platform of platforms) {
> -    const archs = platform === _appBuilderLib().Platform.MAC ? [_builderUtil().Arch.x64] : arch === "all" ? [_builderUtil().Arch.x64, _builderUtil().Arch.ia32] : [(0, _builderUtil().archFromString)(arch == null ? process.arch : arch)];
> +    const archs = arch === "all" && platform !== _appBuilderLib().Platform.MAC ? [_builderUtil().Arch.x64, _builderUtil().Arch.ia32] : [(0, _builderUtil().archFromString)(arch == null ? process.arch : arch)];
>      const archToType = new Map();
>      targets.set(platform, archToType);
>  
> ```

I have Apple Silicon Big Sur machine will check and back to you.

**Comment by ahmadwaliesipick** [comment]
> f (platform === _appBuilderLib().Platform.MAC) {

I have tested but build failed because some packages not found while rebuild

**Comment by ahmadwaliesipick** [comment]
> > f (platform === _appBuilderLib().Platform.MAC) {
> 
> @mmaietta I have tested but build failed because some packages not found while rebuild
> https://github.com/atom/node-keytar/releases/download/v7.0.0/keytar-v7.0.0-electron-v82-darwin-arm64.tar.gz
> 
> https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v3.1.13/node-v64-darwin-x64.tar.gz

@mmaietta here is npm log
https://gist.github.com/ahmadwaliesipick/609d15d82b03d28fbc8a1d788a107766

**Comment by deathlyrage** [comment]
@mmaietta your patches are working I managed to build my app for arm. It can be opened on the Apple DTK but crashes shortly after related to something about a electron, I will investigate and see If I can find any fixes. If you need your app tested on a DTK I'm free to test it.

**Comment by aabuhijleh** [comment]
@abhishekmatcha I expect native node modules are going to cause some issues for arm64 builds (see [blog](https://www.electronjs.org/blog/apple-silicon#potential-issues))

**Comment by aabuhijleh** [comment]
@mmaietta I noticed another place where x64 is hardcoded but it's not included in your patch

[await this.dispatchArtifactCreated(artifactPath, null, Arch.x64, this.computeSafeArtifactName(artifactName, "pkg"))](https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/macPackager.ts#L239)

@deathlyrage can you please also test if you can also sign/notarize your application well?

**Comment by ggreco** [comment]
> @mmaietta I have tested but build failed because some packages not found while rebuild
> https://github.com/atom/node-keytar/releases/download/v7.0.0/keytar-v7.0.0-electron-v82-darwin-arm64.tar.gz

You need at least node-gyp 7.1.2 to build native modules for Mac/ARM64

**Comment by deathlyrage** [comment]
Attempting to use x64 and arm64 targets together result in them uploading under the same name in deployment.
Can be solved by adding  "artifactName" : "${productName}-${arch}-${version}.${ext}", under the mac tag in package.json.

**Comment by deathlyrage** [comment]
@aabuhijleh signing and notarization is working.

**Comment by ahmadwaliesipick** [comment]
> > @mmaietta I have tested but build failed because some packages not found while rebuild
> > https://github.com/atom/node-keytar/releases/download/v7.0.0/keytar-v7.0.0-electron-v82-darwin-arm64.tar.gz
> 
> You need at least node-gyp 7.1.2 to build native modules for Mac/ARM64

I am able to create build and notarize.  But i have sqlite3 error now.
`Uncaught Exception:
Error: dlopen(/Users/kalpit/Documents/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/electron-v11.0-darwin-arm64/node_sqlite3.node, 1): no suitable image found.  Did find:
/Users/kalpit/Documents/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/electron-v11.0-darwin-arm64/node_sqlite3.node: mach-o, but wrong architecture
/Users/kalpit/Documents/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/electron-v11.0-darwin-arm64/node_sqlite3.node: mach-o, but wrong architecture
at process.func [as dlopen] (electron/js2c/asar_bundle.js:5:1812)
at Object.Module._extensions..node (internal/modules/cjs/loader.js:1203:18)
at Object.func [as .node] (electron/js2c/asar_bundle.js:5:2039)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (/Users/kalpit/Documents/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/sqlite3/lib/sqlite3-binding.js:4:15)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (/Users/kalpit/Documents/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/sqlite3/lib/sqlite3.js:2:15)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (/Users/kalpit/Documents/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/any-db-sqlite3/index.js:3:20)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at getAdapter (/Users/kalpit/Documents/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/any-db/index.js:27`

**Comment by ahmadwaliesipick** [comment]
> @aabuhijleh signing and notarization is working. Haven't tested MAS as I'm not using it yet.

can we create dmg for arm64?

**Comment by ahmadwaliesipick** [comment]
I have installed latest sqlite3 but still facing same error
`rror: dlopen(/Users/kalpit/Documents/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node, 1): no suitable image found`

**Comment by deathlyrage** [comment]
> > @aabuhijleh signing and notarization is working. Haven't tested MAS as I'm not using it yet.
> 
> can we create dmg for arm64?

You need to use a dmg or .zip when transfering your .app to your mac, as copying the folder can loose some attributes and macos big sur won't load images that it cant verify, I had this problem and changed the target to dmg and it worked.

**Comment by ljy-xian** [comment]
> I have installed latest sqlite3 but still facing same error
> `rror: dlopen(/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node, 1): no suitable image found`

I 'm not good at english ~
try lipo -info "your node_sqlite3.node path" see which arch it is 
sqlite3 is using node-pre-gyp to build addon , node-pre-gyp  is not support apple silicon , when you set "--arch=arm64" as configuration it don't  work
you can try electron-rebuild to rebuild your addon , then try electron-builder to build the dmg
i'm trying this, it seems work

**Comment by mmaietta** [comment]
Wow, I wasn't expecting such quick responses! 😁 

> @mmaietta I have tested but build failed because some packages not found while rebuild
https://github.com/atom/node-keytar/releases/download/v7.0.0/keytar-v7.0.0-electron-v82-darwin-arm64.tar.gz
https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v3.1.13/node-v64-darwin-x64.tar.gz

@ahmadwaliesipick you'll need 11.0.0-beta.1 or higher, that's when arm64 dist started to be released for Electron. For any native modules, you'll likely need to recompile directly as @ljy-xian mentioned since it'll be highly unlikely any other packages have already been packaging/storing arm64 builds for mac yet. The sqlite error is due to the fact that your native addon is not available in arm64 - `no suitable image found`

From: https://www.electronjs.org/blog/apple-silicon#native-modules
You need the following dependencies + versions
- Xcode	>=12.0.0
- node-gyp	>=7.1.0
- electron-rebuild	>=1.12.0
- electron-packager	

>Can be solved by adding "artifactName" : "${productName}-${arch}-${version}.${ext}", under the mac tag in package.json.

I recall some additional code that was conditional for adding the arch in the name, but I'll need to look that back up.  I previously had to write some scripting for doing the same artifactName arch-injection for compiling ia32 vs x64 Windows apps separately.  Seems the patch files might need to get a bit longer, possibly better in a gist soon

Nice catch @aabuhijleh! I don't use mas and definitely overlooked that point. My patch was to kickstart towards what could be a worthwhile PR.

**Comment by ahmadwaliesipick** [comment]
> > I have installed latest sqlite3 but still facing same error
> > `rror: dlopen(/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node, 1): no suitable image found`
> 
> I 'm not good at english ~
> try lipo -info "your node_sqlite3.node path" see which arch it is
> sqlite3 is using node-pre-gyp to build addon , node-pre-gyp is not support apple silicon , when you set "--arch=arm64" as configuration it don't work
> you can try electron-rebuild to rebuild your addon , then try electron-builder to build the dmg
> i'm trying this, it seems work

@ljy-xian Thanks for reply.
i am already using electron-rebuild.
can you verify the command and target should be this?
`electron-builder  --arm64`
`"target" : ["dmg","pkg"],`

**Comment by ahmadwaliesipick** [comment]
> Wow, I wasn't expecting such quick responses! 😁
> 
> > @mmaietta I have tested but build failed because some packages not found while rebuild
> > https://github.com/atom/node-keytar/releases/download/v7.0.0/keytar-v7.0.0-electron-v82-darwin-arm64.tar.gz
> > https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v3.1.13/node-v64-darwin-x64.tar.gz
> 
> @ahmadwaliesipick you'll need 11.0.0-beta.1 or higher, that's when arm64 dist started to be released for Electron. For any native modules, you'll likely need to recompile directly as @ljy-xian mentioned since it'll be highly unlikely any other packages have already been packaging/storing arm64 builds for mac yet. The sqlite error is due to the fact that your native addon is not available in arm64 - `no suitable image found`
> 
> From: https://www.electronjs.org/blog/apple-silicon#native-modules
> You need the following dependencies + versions
> 
> * Xcode	>=12.0.0
> * node-gyp	>=7.1.0
> * electron-rebuild	>=1.12.0
> * electron-packager	>=15.1.0
> 
> > Can be solved by adding "artifactName" : "${productName}-${arch}-${version}.${ext}", under the mac tag in package.json.
> 
> I recall some additional code that was conditional for adding the arch in the name, but I'll need to look that back up. I previously had to write some scripting for doing the same artifactName arch-injection for compiling ia32 vs x64 Windows apps separately. Seems the patch files might need to get a bit longer, possibly better in a gist soon
> 
> Nice catch @aabuhijleh! I don't use mas and definitely overlooked that point.

@mmaietta  i am using 11.0.0-beta.20

**Comment by ljy-xian** [comment]
> erify the command and target should be this?

i path electron-builder directly as @mmaietta  mentioned
then run electron-builder --arm64 it works
then run the dmg on silicon MacOS 11.0.1 beta , it works too

**Comment by ahmadwaliesipick** [comment]
> > erify the command and target should be this?
> 
> i patch electron-builder directly as @mmaietta mentioned
> then run electron-builder --arm64 it works
> then run the dmg on silicon MacOS 11.0.1 beta , it works too

@ljy-xian  what should be in target?

**Comment by ahmadwaliesipick** [comment]
Please anyone share the correct format of electron-builder params and target params to create arm64 dmg.

**Comment by ljy-xian** [comment]
> Please anyone share the correct format of electron-builder params and target params to create arm64 dmg.
i use electron cli to build ， however it's the same
i think electron-builder will build dmg as default
./.node/bin/node ./node_modules/electron-builder/out/cli/cli.js --arm64 --config.productName=微盘
it will build both mac and dmg
![image](https://user-images.githubusercontent.com/10915655/99036406-92171e00-25bc-11eb-8092-c4ce1dff20e9.png)

**Comment by ahmadwaliesipick** [comment]
> > Please anyone share the correct format of electron-builder params and target params to create arm64 dmg.
> > i use electron cli to build ， however it's the same
> > i think electron-builder will build dmg as default
> > ./.node/bin/node ./node_modules/electron-builder/out/cli/cli.js --arm64 --config.productName=微盘
> > it will build both mac and dmg
> > ![image](https://user-images.githubusercontent.com/10915655/99036406-92171e00-25bc-11eb-8092-c4ce1dff20e9.png)

trying again i have missed patch :-D

**Comment by ahmadwaliesipick** [comment]
> @abhishekmatcha
> i use electron cli to build ， however it's the same
> i think electron-builder will build dmg as default
> ./.node/bin/node ./node_modules/electron-builder/out/cli/cli.js --arm64 --config.productName=微盘
> it will build both mac and dmg
> ![image](https://user-images.githubusercontent.com/10915655/99036406-92171e00-25bc-11eb-8092-c4ce1dff20e9.png)

@ljy-xian are you also using sqlite3? do you have issue with that?

**Comment by ahmadwaliesipick** [comment]
facing one more issue while creating dmg.

` command=/Users/kalpit/Documents/project/todo-xplat/node_modules/app-builder-bin/mac/app-builder blockmap --input /Users/kalpit/Documents/project/todo-xplat/dist-app/Todo Cloud-TEST-10.5.11-mac.dmg --output /Users/kalpit/Documents/project/todo-xplat/dist-app/Todo Cloud-TEST-10.5.11-mac.dmg.blockmap
assertion failed [abi_info.kind == AbiKind::TranslatedCode]: emulated forward to an arm pc that isn't in translated code. arm_pc=0x10242d4e8 abi_kind=6 emulation_interval=[0x102613234,0x10261324c) instruction_interval=[0x102613230, 0x10261324c) x86_rip=0x1436ff9
(ThreadContextRegisterState.cpp:677 move_to_instruction_boundary)`

**Comment by ljy-xian** [comment]
@ahmadwaliesipick  
@ljy-xian are you also using sqlite3? do you have issue with that?
yes , but it seem works well
by the way , can you please delete my product's snapshots in your replies :-D

**Comment by ahmadwaliesipick** [comment]
> @ahmadwaliesipick
> @ljy-xian are you also using sqlite3? do you have issue with that?
> yes , but it seem works well
> by the way , can you please delete my product's snapshots in your replies :-D

@ljy-xian I have removed snapshots

Which sqlite3 version you are using?

**Comment by ahmadwaliesipick** [comment]
```
Uncaught Exception:
Error: dlopen(/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node, 1): no suitable image found.  Did find:
/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node: mach-o, but wrong architecture
/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node: mach-o, but wrong architecture
at process.func [as dlopen] (electron/js2c/asar_bundle.js:5:1812)
at Object.Module._extensions..node (internal/modules/cjs/loader.js:1203:18)
at Object.func [as .node] (electron/js2c/asar_bundle.js:5:2039)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/sqlite3/lib/sqlite3-binding.js:4:15)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/sqlite3/lib/sqlite3.js:2:15)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/any-db-sqlite3/index.js:3:20)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at getAdapter (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/any-db/index.js:27:10)
at Object.connect [as createConnection] (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/any-db/index.js:14:17)
```

**Comment by ljy-xian** [comment]
> > @ahmadwaliesipick
> > @ljy-xian are you also using sqlite3? do you have issue with that?
> > yes , but it seem works well
> > by the way , can you please delete my product's snapshots in your replies :-D
> 
> @ljy-xian I have removed snapshots
> 
> Which sqlite3 version you are using?

3.31.1 
did you use lipo -info  command to test your addon in node-sqlite3 ?

**Comment by ahmadwaliesipick** [comment]
> did you use lipo -info command to test your addon in node-sqlite3 ?

I have no idea about this lipo command. can you please me correct format if this command?

**Comment by deathlyrage** [comment]
Read up on lipo here: https://developer.apple.com/documentation/xcode/building_a_universal_macos_binary

**Comment by Eugeny** [comment]
@mmaietta's patches work perfectly - except for native deps. I've set `"npmRebuild": false,` in package.json and used electron-rebuild with `--arch arm64` to pre-build them

**Comment by warpdesign** [comment]
@mmaietta how about making a PR with these changes?

**Comment by ahmadwaliesipick** [comment]
Guys i am stuck with sqlite3 rebuild issue: 
`   node-pre-gyp WARN Using request for node-pre-gyp https download 
    node-pre-gyp http GET https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v4.2.0/electron-v11.0-darwin-arm64.tar.gz
    node-pre-gyp http 403 https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v4.2.0/electron-v11.0-darwin-arm64.tar.gz
    node-pre-gyp WARN Tried to download(403): https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v4.2.0/electron-v11.0-darwin-arm64.tar.gz 
    node-pre-gyp WARN Pre-built binaries not found for sqlite3@4.2.0 and electron@11.0.0-beta.23 (electron-v11.0 ABI, unknown) (falling back to source compile with node-gyp) 
    node-pre-gyp http 403 status code downloading tarball https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v4.2.0/electron-v11.0-darwin-arm64.tar.gz 
    node-pre-gyp verb command build [ 'rebuild' ]
`

anyone please help.

**Comment by ahmadwaliesipick** [comment]
> > > @ahmadwaliesipick
> > > @ljy-xian are you also using sqlite3? do you have issue with that?
> > > yes , but it seem works well
> > > by the way , can you please delete my product's snapshots in your replies :-D
> > 
> > 
> > @ljy-xian I have removed snapshots
> > Which sqlite3 version you are using?
> 
> 3.31.1
> did you use lipo -info command to test your addon in node-sqlite3 ?

@ljy-xian its showing x86_64 :-( 
```
 lipo -archs  dist-app/mac-arm64/Todo\ Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node 

x86_64
```

**Comment by ahmadwaliesipick** [comment]
> 3.31.1

@ljy-xian sqlite3 version 3.31.1 not exists can you please confirm again?

**Comment by mmaietta** [comment]
> @mmaietta how about making a PR with these changes?

@warpdesign, happy to attempt to. I'll need others to test it on their devices (DTK and/or BigSur) since I can't test locally on my company laptop. I finally was able to get tests on `master` branch to work locally for me today, so I'll aim to work on a PR this week.

**Comment by deathlyrage** [comment]
@mmaietta I'm able to test for you if you want to message me with any details needed. Have both DTK and several silicon devices shipping in the first launch window.

**Comment by warpdesign** [comment]
I can do some testing as well if needed (BigSur, x64).

**Comment by mmaietta** [comment]
Opened PR https://github.com/electron-userland/electron-builder/pull/5426
Added instructions on how to set up a local dev environment from scratch with my feature branch and the electron-quick-start repo.  You should also be able to apply the same instructions (`yalc link <all the electron-builder packages>`) for testing this with your own project as well.

**Comment by mmaietta** [comment]
As it turns out, my PR also enables you to use `install-app-deps` and `buildDependenciesFromSource` successfully for multiple archs.  I added sqlite3 to the package.json and node-gyp builds with the correct architecture.

Example output:
```
$ electron-builder
  • electron-builder  version=22.8.1 os=19.6.0
  • loaded configuration  file=package.json ("build" field)
  • writing effective config  file=dist/builder-effective-config.yaml
  • rebuilding native dependencies  dependencies=sqlite3@5.0.0 platform=darwin arch=arm64
  • rebuilding native dependency  name=sqlite3 version=5.0.0
  • packaging       platform=darwin arch=arm64 electron=11.0.0 appOutDir=dist/mac-arm64
  • default Electron icon is used  reason=application icon is not set
  • skipped macOS code signing  reason=identity explicitly is set to null
  • building        target=DMG arch=arm64 file=dist/electron-quick-start-1.0.0-arm64.dmg
  • rebuilding native dependencies  dependencies=sqlite3@5.0.0 platform=darwin arch=x64
  • rebuilding native dependency  name=sqlite3 version=5.0.0
  • building block map  blockMapFile=dist/electron-quick-start-1.0.0-arm64.dmg.blockmap
  • packaging       platform=darwin arch=x64 electron=11.0.0 appOutDir=dist/mac
  • skipped macOS code signing  reason=identity explicitly is set to null
  • building        target=DMG arch=x64 file=dist/electron-quick-start-1.0.0.dmg
  • building block map  blockMapFile=dist/electron-quick-start-1.0.0.dmg.blockmap
✨  Done in 157.86s.
```

Could anyone please test running these two builds for me on Big Sur (both x64 and DTK)? 
[electron-quick-start-1.0.0-arm64.dmg](https://drive.google.com/file/d/1XM8gZe_Fo5SrciDVMo7elc5keGBY4E5k/view?usp=sharing)
[electron-quick-start-1.0.0.dmg](https://drive.google.com/file/d/1dFzlno83Ns8zXM_8chE0zd_AQ-vqiJdz/view?usp=sharing)
Notes: It'll be a blank screen, it's just electron's example app. I presume x64 app would work with Rosetta 2, but IIRC that's not available to test on DTKs.

**Comment by ahmadwaliesipick** [comment]
> I added sqlite3 to the package.json and node-gyp builds with the correct architecture.

@mmaietta can you please share package.json file?

**Comment by mmaietta** [comment]
> @mmaietta can you please share package.json file?

```
{
  "name": "electron-quick-start",
  "version": "1.0.0",
  "description": "A minimal Electron application",
  "main": "main.js",
  "scripts": {
    "start": "electron .",
    "build": "electron-builder"
  },
  "repository": "https://github.com/electron/electron-quick-start",
  "keywords": [
    "Electron",
    "quick",
    "start",
    "tutorial",
    "demo"
  ],
  "author": "GitHub",
  "license": "CC0-1.0",
  "dependencies": {
    "sqlite3": "^5.0.0"
  },
  "devDependencies": {
    "electron": "^11.0.0",
    "electron-builder": "^22.9.1"
  },
  "build": {
    "buildDependenciesFromSource": true,
    "mac": {
      "identity": null,
      "target": {
        "target": "dmg",
        "arch": [
          "arm64",
          "x64"
        ]
      }
    }
  }
}

```

**Comment by ahmadwaliesipick** [comment]
@mmaietta  can you please share which node and node-gyp version you are using?

**Comment by mmaietta** [comment]
> @mmaietta can you please share which node and node-gyp version you are using?

```
$ node -v
v12.16.1
$ npm -version
6.13.4
```
Excerpt from `package-lock.json`
```
"node-gyp": {
      "version": "3.8.0",
      "resolved": "https://registry.npmjs.org/node-gyp/-/node-gyp-3.8.0.tgz",
```

**Comment by ahmadwaliesipick** [comment]
> > @mmaietta can you please share which node and node-gyp version you are using?
> 
> ```
> $ node -v
> v12.16.1
> $ npm -version
> 6.13.4
> ```
> 
> Excerpt from `package-lock.json`
> 
> ```
> "node-gyp": {
>       "version": "3.8.0",
>       "resolved": "https://registry.npmjs.org/node-gyp/-/node-gyp-3.8.0.tgz",
> ```

but as per https://www.electronjs.org/blog/apple-silicon `node-gyp` must be` >=7.1.0`

**Comment by mmaietta** [comment]
¯\_(ツ)_/¯

You're correct, I don't know how it compiled then

If you're willing to set up your local env to test out the feature branch (instructions in the PR), that'd help immensely with debugging this.

**Comment by ahmadwaliesipick** [comment]
> ¯\_(ツ)_/¯
> 
> You're correct, I don't know how it compiled then
> 
> If you're willing to set up your local env to test out the feature branch (instructions in the PR), that'd help immensely with debugging this.

if i used node-gyp >=7.1.0  then the build showing error. can you please share feature branch name?

**Comment by mmaietta** [comment]
PR should have all the info you need... https://github.com/electron-userland/electron-builder/pull/5426
Fork -> mmaietta:Feature/AppleSiliconARM64

**Comment by ahmadwaliesipick** [comment]
@mmaietta i have used above fork but i am still facing sqlite error :-(
`Uncaught Exception:
Error: dlopen(project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node, 1): no suitable image found.  Did find:
/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node: mach-o, but wrong architecture
/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node: mach-o, but wrong architecture
at process.func [as dlopen] (electron/js2c/asar_bundle.js:5:1812)
at Object.Module._extensions..node (internal/modules/cjs/loader.js:1203:18)
at Object.func [as .node] (electron/js2c/asar_bundle.js:5:2039)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/sqlite3/lib/sqlite3-binding.js:4:15)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/sqlite3/lib/sqlite3.js:2:15)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at Object.<anonymous> (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/any-db-sqlite3/index.js:3:20)
at Module._compile (internal/modules/cjs/loader.js:1152:30)
at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
at Module.load (internal/modules/cjs/loader.js:992:32)
at Module._load (internal/modules/cjs/loader.js:885:14)
at Function.f._load (electron/js2c/asar_bundle.js:5:12694)
at Module.require (internal/modules/cjs/loader.js:1032:19)
at require (internal/modules/cjs/helpers.js:72:18)
at getAdapter (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/any-db/index.js:27:10)
at Object.connect [as createConnection] (/project/todo-xplat/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar/node_modules/todo-api/node_modules/any-db/index.js:14:17)`

**Comment by ahmadwaliesipick** [comment]
@mmaietta i think i have figured out i have issue with `std `version https://github.com/electron/electron/issues/26364 
can you please confirm what is `gcc ` version that you are using?

**Comment by mmaietta** [comment]
Yes, my team was able to resolve our native addon compilation issue by changing our cflags from `-std=c++11` to `c++14`

As mentioned previously, I am not able to have Big Sur installed (at this time).

**Comment by mmaietta** [comment]
I can confirm that x64 dmg + app launches on Big Sur. 
Can anyone please test this arm64 build on a DTK?
[electron-quick-start-1.0.0-arm64.dmg](https://drive.google.com/file/d/1XM8gZe_Fo5SrciDVMo7elc5keGBY4E5k/view?usp=sharing)

**Comment by deathlyrage** [comment]
'electron-quick-start.app' is damaged and can't be opened. You should move it to the bin. (ON DTK)
Even after disabling and allowing apps from anywhere. Is it possible to get this as signed under a Developer ID cert so I don't have to resign it and possibly ruin the test results ^

Keep in mind big sur requires all apps to be signed now.

**Comment by mmaietta** [comment]
@develar @warpdesign It seems that `master` branch is broken for both builds and tests:
https://app.circleci.com/pipelines/github/electron-userland/electron-builder/172/workflows/8e811c33-4c20-41ff-affc-d4a6a7c3d71a/jobs/3169
This PR resolves that compilation issue and updates a few snapshots https://github.com/electron-userland/electron-builder/pull/5434. Please review when you have a chance

@deathlyrage my apologies, I overlooked that tid bit. I've reuploaded Dev-signed artifacts, same google drive url. (`identityName=Apple Development: <email> (24KMVXPDMX)`)

**Comment by ahmadwaliesipick** [comment]
> electron-quick-start-1.0.0-arm64.dmg
@mmaietta  I have installed and run app at Apple Silicon BigSur machine. here is screenshot
https://share.getcloudapp.com/RBu96Zvb

**Comment by mmaietta** [comment]
For anyone running into this issue with building arm64 native addons with node-gyp 7.1.0, you're likely receiving "Unsupported architecture" errors akin to this below (Note: `MacOSX10.15.sdk`)
```
    In file included from Release/obj/gen/sqlite-autoconf-3310100/sqlite3.c:14012:
    In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/stdio.h:64:
    In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/_stdio.h:68:
    /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
    #error Unsupported architecture
     ^
```
The solution is to get the MacOSX 11.1 SDK, which contains the necessary files for compiling for both arm64 and x64, which isn't available in Xcode 12.2. Steps below:

1. Install Xcode Command Line Tools 12.3 Beta
2. `export SDKROOT="/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk"`
3. `yarn electron-builder install-app-deps --arch arm64`

If still having trouble, try switching to python 3.8, and you can force it via `export NODE_GYP_FORCE_PYTHON=<path>`

Still working on resolving the remaining PR unit tests.

Side note:
For anyone that finds the node-gyp output via electron-builder logs obnoxious, here's a patch that cleans up the command wrapper. Basically, `error.stack` is `error.message` but with way too many CRLF, so the logging gets duplicated. This significantly helped me debugging the logs.
`electron-builder+22.9.1.patch`
```
diff --git a/node_modules/electron-builder/out/cli/cli.js b/node_modules/electron-builder/out/cli/cli.js
index e69d729..bab7423 100755
--- a/node_modules/electron-builder/out/cli/cli.js
+++ b/node_modules/electron-builder/out/cli/cli.js
@@ -160,9 +160,7 @@ function wrap(task) {
       if (error instanceof _builderUtil().InvalidConfigurationError) {
         _builderUtil().log.error(null, error.message);
       } else if (!(error instanceof _util().ExecError) || !error.alreadyLogged) {
-        _builderUtil().log.error({
-          stackTrace: error.stack
-        }, error.message);
+        _builderUtil().log.error(null, error.message);
       }
     });
   };
```

**Comment by ahmadwaliesipick** [comment]
> For anyone running into this issue with building arm64 native addons with node-gyp 7.1.0, you're likely receiving "Unsupported architecture" errors akin to this below (Note: `MacOSX10.15.sdk`)
> 
> ```
>     In file included from Release/obj/gen/sqlite-autoconf-3310100/sqlite3.c:14012:
>     In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/stdio.h:64:
>     In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/_stdio.h:68:
>     /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
>     #error Unsupported architecture
>      ^
> ```
> 
> The solution is to get the MacOSX 11.1 SDK, which contains the necessary files for compiling for both arm64 and x64, which isn't available in Xcode 12.2. Steps below:
> 
> 1. Install Xcode Command Line Tools 12.3 Beta
> 2. `export SDKROOT="/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk"`
> 3. `yarn electron-builder install-app-deps --arch arm64`
> 
> If still having trouble, try switching to python 3.8, and you can force it via `export NODE_GYP_FORCE_PYTHON=<path>`
> 
> Still working on resolving the remaining PR unit tests.
> 
> Side note:
> For anyone that finds the node-gyp output via electron-builder logs obnoxious, here's a patch that cleans up the command wrapper. Basically, `error.stack` is `error.message` but with way too many CRLF, so the logging gets duplicated. This significantly helped me debugging the logs.
> `electron-builder+22.9.1.patch`
> 
> ```
> diff --git a/node_modules/electron-builder/out/cli/cli.js b/node_modules/electron-builder/out/cli/cli.js
> index e69d729..bab7423 100755
> --- a/node_modules/electron-builder/out/cli/cli.js
> +++ b/node_modules/electron-builder/out/cli/cli.js
> @@ -160,9 +160,7 @@ function wrap(task) {
>        if (error instanceof _builderUtil().InvalidConfigurationError) {
>          _builderUtil().log.error(null, error.message);
>        } else if (!(error instanceof _util().ExecError) || !error.alreadyLogged) {
> -        _builderUtil().log.error({
> -          stackTrace: error.stack
> -        }, error.message);
> +        _builderUtil().log.error(null, error.message);
>        }
>      });
>    };
> ```

@mmaietta  this also works for `node-sqlite3`  https://github.com/mapbox/node-sqlite3 ?

**Comment by ggreco** [comment]
> The solution is to get the MacOSX 11.1 SDK, which contains the necessary files for compiling for both arm64 and x64, which isn't available in Xcode 12.2. Steps below:

The full install of Xcode 12.2 includes the MacOSX 11 SDK, the problem is that the command line tools do not.

I suggest not to mix commandlinetools and xcode full installs since this usually causes a lot of problems.

Thanks for your work @mmaietta !

Now that electron 11 is out of the beta stage electron-builder is the last piece of the puzzle to be able to ship in a "clean" way our electron app also for M1 users!

**Comment by warpdesign** [comment]
@mmaietta I was able to build both x64 & ARM64 .dmg on my project using your fork/branch.

Nice job!

I could only test the x64 version though. Could anyone confirm the [ARM64](https://www.warpdesign.fr/tests/arm64/React-Explorer-2.3.1-arm64.dmg) version is working?

**Comment by gchilds** [comment]
> I could only test the x64 version though. Could anyone confirm the [ARM64](https://www.warpdesign.fr/tests/arm64/React-Explorer-2.3.1-arm64.dmg) version is working?

I get this when I try to run it
![image](https://user-images.githubusercontent.com/2145453/100137589-f8d5ea80-2e8c-11eb-8e42-f1ab5579a14f.png)

**Comment by mmaietta** [comment]
Working on fixing unit tests as the tests needed to upgrade to use electron 11, need to set up a windows VM to fix the remaining tests and test on another mac as my company laptop basically quarantines any dmg hdiutil creates 😅 

Also worth noting, it seems that Big Sur removes hdiutil's flatten/unflatten, so that will need to be updated at some point, but I'd consider it outside the scope of this PR though since this builds correctly on Catalina.
For Big Sur we may need to investigate migrating to hdiutil's:
```
udifrez [options] image
                embed resources (e.g. a software license agreement) in a disk image.
udifderez [options] image
                extract resources from image.
```

**Comment by mmaietta** [comment]
@warpdesign I don't see `node-gyp` in the package-lock.json, I'm not sure what it was using to compile your native addon.
I cloned your repo, installed node-gyp, added `arm64` and `buildDependenciesFromSource` to package.json mac target, forced MacOSX 11.1 sdk, and then recompiled + dev signed it. These are the builds if someone can test launching it on a DTK
[React-Explorer-2.3.1-arm64.dmg](https://drive.google.com/file/d/188QU7Wd7GjNMeusY8mlgHNA_LYZEF__Y/view?usp=sharing)
[React-Explorer-2.3.1.dmg](https://drive.google.com/file/d/1DPagNj29_mxG0nzK1M1xPFbq_VZhSRAn/view?usp=sharing)

**Comment by tmacwill** [comment]
Thanks for your work on this @mmaietta! I'm working on an electron-builder app and I have an M1 MacBook Pro running Big Sur, so I'm happy to help out.

I tried both of those DMGs, and they had the same behavior for me:

- I got a security warning on launch, but opened the app via the Security & Privacy panel.
- The app just shows this—not sure what the intended behavior is, though.

<img width="912" alt="Screen Shot 2020-11-24 at 7 10 03 PM" src="https://user-images.githubusercontent.com/88533/100178595-05584400-2e89-11eb-9a37-ad0bdb50ad11.png">

- I'm unable to quit the app. None of the close window button, ⌘-W, ⌘-Q, or menu bar options work. I had to force quit the app. Not sure if that's due to react-explorer or electron-builder.

More than happy to test more if it'd be helpful!

**Comment by warpdesign** [comment]
@tmacwill thanks for the tests!

Could you choose "Toggle Developer Tools" from the Window menu, click on "console", and copy/paste what is displayed here please?

I have created an issue here: https://github.com/warpdesign/react-explorer/issues/197

**Comment by warpdesign** [comment]
@mmaietta no idea why node-gyp isn't there. Native dependencies are supposed to be handled by electron-rebuild which is installed/used by electron-builder. As far as I know, I'm not supposed to include it myself in my project.

The builds you made don't seem to have the native dependency in the asar archive (`drivelist.node`): this explains why it's stuck on the loading screen. That's weird.

**Comment by tmacwill** [comment]
Sure, screenshot below. If this is specific to react-explorer, the issue you created is a better place to discuss, but if this is relevant to electron-builder itself, we can keep going here.

<img width="1552" alt="Screen Shot 2020-11-25 at 12 20 53 AM" src="https://user-images.githubusercontent.com/88533/100200900-24200000-2eb4-11eb-9732-e7ff825280c7.png">

**Comment by warpdesign** [comment]
> @warpdesign I don't see `node-gyp` in the package-lock.json, I'm not sure what it was using to compile your native addon.
> I cloned your repo, installed node-gyp, added `arm64` and `buildDependenciesFromSource` to package.json mac target, forced MacOSX 11.1 sdk, and then recompiled + dev signed it. These are the builds if someone can test launching it on a DTK
> [React-Explorer-2.3.1-arm64.dmg](https://drive.google.com/file/d/188QU7Wd7GjNMeusY8mlgHNA_LYZEF__Y/view?usp=sharing)
> [React-Explorer-2.3.1.dmg](https://drive.google.com/file/d/1DPagNj29_mxG0nzK1M1xPFbq_VZhSRAn/view?usp=sharing)

I downloaded and extracted the DMG of the ARM64 version and the `drivelist.node` file in `Contents/ressource/app.asar/node_modules/drivelist/build/Release/drivelist.node` is an x64 binary.

It seems native dependencies are built for the host architecture, and not the specified --arch.

**Comment by ggreco** [comment]
About the workflow in electron-builder and native modules.

I use a module that is prebuilt for the different architectures, at the moment in the "build" settings in package.json I have this:

```json
   "mac": {
      "category": "public.app-category.sports",
      "extraResources": [
        {
          "from": "node_modules/@wyscout/hudl-graphx-core/prebuilds/darwin-x64",
          "to": "app/node_modules/@wyscout/hudl-graphx-core/prebuilds/darwin-x64",
          "filter": [ "**/*"  ]
        }
      ]
    },
    "win": {
      "extraResources": [
        {
          "from": "node_modules/@wyscout/hudl-graphx-core/prebuilds/win32-x64",
          "to": "app/node_modules/@wyscout/hudl-graphx-core/prebuilds/win32-x64",
          "filter": [ "**/*" ]
        }
      ]
    }
```

The arch switch you added in the "mac" section should allow me to write also:

```
 "from": "node_modules/@wyscout/hudl-graphx-core/prebuilds/darwin-${arch}"
 "to": "app/node_modules/@wyscout/hudl-graphx-core/prebuilds/darwin-${arch}"
```

... correct?

**Comment by warpdesign** [comment]
@tmacwill could you try downloading both [arm64](https://www.warpdesign.fr/tests/arm64/React-Explorer-2.3.1-arm64.dmg) & [x64](https://www.warpdesign.fr/tests/arm64/React-Explorer-2.3.1.dmg) builds again?

**Comment by tmacwill** [comment]
getting the same security error, but otherwise the app looks like it works now (and also quits correctly).

**Comment by warpdesign** [comment]
> getting the same security error, but otherwise the app looks like it works now (and also quits correctly).

Good news! Which version (arm, x64) did you try?

**Comment by tmacwill** [comment]
Both, same behavior.

**Comment by warpdesign** [comment]
Great!

Could you show me the about window of both versions please?

Le mer. 25 nov. 2020 à 19:11, Tommy MacWilliam <notifications@github.com> a
écrit :

**Comment by mmaietta** [comment]
@warpdesign I noticed that for your node module, `driver`, it only compiles an x64 even when node-gyp is executed with arm64. Oddly, `sqlite3` generates both archs and I'm not sure why there's differing behavior here:
```
./node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node
./node_modules/sqlite3/lib/binding/napi-v3-darwin-x64/node_sqlite3.node

./node_modules/drivelist/build/Release/drivelist.node
```

**Comment by warpdesign** [comment]
Not sure what happens but I successfully built an arm64 binary using
`electron-rebuild --arch arm64`.

Le mer. 25 nov. 2020 à 19:41, Mike Maietta <notifications@github.com> a
écrit :

**Comment by tmacwill** [comment]
<img width="912" alt="Screen Shot 2020-11-25 at 11 44 54 AM" src="https://user-images.githubusercontent.com/88533/100274734-cfac6d00-2f13-11eb-8cd9-8bf508dddc92.png">

<img width="912" alt="Screen Shot 2020-11-25 at 11 45 40 AM" src="https://user-images.githubusercontent.com/88533/100274748-d3d88a80-2f13-11eb-9835-bfc8d70b36b8.png">

**Comment by mmaietta** [comment]
@warpdesign what happens when you attempt to build a native addon for x64 and arm64 in the same pipeline?
AFAICT, node-gyp doesn't have an output path configuration so only one artifact appears at /build/Release, right?

**Comment by mmaietta** [comment]
I've updated my arm64-support PR but the CI can't give its stamp of approval because `master` branch is already broken. The PRs being merged aren't showing build/tests passing, and master branch naturally doesn't either. The build literally can't complete, and also worrisome, yarn.lock wasn't updated properly to include the new npm dependency it added.

@develar I've started another PR. It currently fixes the compiler issues, but it seems more tests are still broken. I'm investigating further. https://github.com/electron-userland/electron-builder/pull/5449

**Comment by mmaietta** [comment]
Quick updates:

Compiler PR https://github.com/electron-userland/electron-builder/pull/5449 fixing master branch is approved and ready for merge.

Apple Silicon PR https://github.com/electron-userland/electron-builder/pull/5426 is approved with PR tests passing (CI post-merge test is failing and fixed in Compiler PR linked above). Once the compiler PR is merged, I would advocate for rerunning the CI builds for the Apple Silicon PR for proper validation against master branch and just as a merge strategy, in general. We should then be ready for a pre-release cut for broader testing.
Note: Universal x64+arm64 Electron builds were not available at the time and has not been tested yet.

**Comment by develar** [comment]
Thanks a lot for your work. I will take a look and release a pre-release version for test on this weekend.

**Comment by mmaietta** [comment]
Thanks @develar!

Quick note to all on universal binaries, electron has a package that'll stitch the two .apps together. I haven't tried it out yet, but the process seems fairly straightforward. It's worth noting that it unsurprisingly doubles the app size.
https://github.com/electron/universal

1. `electron-builder -m --dir --arch=x64`
2. `node-gyp clean`
3. `electron-builder -m --dir --arch=arm64`
4. `node <script_using_'makeUniversalApp'>`

That of course can be consolidated into a single js/ts script using electron-builders programmatic API, as presumably you already have some script used for signing+notarization, so why not add everything in one? 
Something like this I'd imagine.
```
import { makeUniversalApp } from '@electron/universal';
import * as builder from "electron-builder";
import { notarize } from "electron-notarize";
import { promisify } from "util";
import webpack from "webpack";
import baseConfig from "../electron-builder.js";
import commonConfig from "../webpack.config.ts";
import { execSync } from "child_process";

let result = (async () => {
  await promisify(webpack)(commonConfig);

  const build = async (arch) => {
    const targets = builder.createTargets([builder.Platform.MAC], 'dir', arch)
    const config: builder.Configuration = {
      ...baseConfig,
      afterSign: async (context: builder.AfterPackContext) => {
        if (context.electronPlatformName === "darwin") {
          await notarize(...)
        }
      },
    }
    return builder.build({ targets, config })
  }

  await build('x64')
  execSync('yarn node-gyp clean');
  await build('arm64')

  await makeUniversalApp({
    x64AppPath: 'dist/mac/Electron.app',
    arm64AppPath: 'dist/mac-arm64/Electron.app',
    outAppPath: 'dist/mac-universal/Electron.app',
  });
})()
```
Note: that's untested, kind of threw it together in text edit with some copy-paste

**Comment by Slidemagic** [comment]
> Also worth noting, it seems that Big Sur removes hdiutil's flatten/unflatten, so that will need to be updated at some point, but I'd consider it outside the scope of this PR though since this builds correctly on Catalina.
> For Big Sur we may need to investigate migrating to hdiutil's:
> 
> ```
> udifrez [options] image
>                 embed resources (e.g. a software license agreement) in a disk image.
> udifderez [options] image
>                 extract resources from image.
> ```

This now generates an error when adding a software license: **⨯ Exit code: 1. Command failed: hdiutil unflatten *******.dmg
hdiutil: unflatten: verb not recognized**

BigSur 11.0.1
Electron 11.0.4
Electron-builder 22.9.1

**Comment by mmaietta** [comment]
@Slidemagic, that should be fixed in this PR https://github.com/electron-userland/electron-builder/pull/5424 and the compilation errors it introduced are fixed in this PR that was just merged: https://github.com/electron-userland/electron-builder/pull/5449
We need a pre-release so you can pull those changes

**Comment by ahmadwaliesipick** [comment]
@mmaietta  i am still not able to resolve `sqlite3 `error. Do you have any solution?

> I have installed latest sqlite3 but still facing same error
> `rror: dlopen(/dist-app/mac-arm64/Todo Cloud.app/Contents/Resources/app.asar.unpacked/node_modules/todo-api/node_modules/sqlite3/lib/binding/napi-v3-darwin-arm64/node_sqlite3.node, 1): no suitable image found`

**Comment by mmaietta** [comment]
@ahmadwaliesipick because you're getting the same error for the same issue, and the info you're providing is very limited to assist.
Your bindings npm module can't find the correct file, likely because it's not available/copied into your `.app`. My advice, check your app.asar.unpacked and see what node images are actually present. To simplify your work, just run this in the working directory: `find . -name '*.node' | xargs -I {} lipo -info {} | grep -e Non-fat`

Testing with `electron-rebuild -f -o sqlite3 --arch=arm64` displays errors, so I assume the project owners need to update it to support the macOS 11 SDK. Excerpt:
```
⠸ Building module: sqlite3, Completed: 0 shell-builder/node_modules/node-addon-api/napi.h:2083:25: error: unknown type name
      'napi_threadsafe_function_call_mode'
                        napi_threadsafe_function_call_mode mode) const;
```

For reference, here's a successful build of leveldown:
```
$ rm -rf node_modules/**/*.node; yarn node-gyp clean;
  electron-rebuild -f -o leveldown --arch arm64 && \
  (find . -name '*.node' | xargs -I {} lipo -info {} | grep -e Non-fat)

Non-fat file: ./node_modules/leveldown/bin/darwin-arm64-69/leveldown.node is architecture: arm64
Non-fat file: ./node_modules/leveldown/build/Release/leveldown.node is architecture: arm64
```

**Comment by mmaietta** [comment]
PR has been merged and release cut performed! Please upgrade and begin stress testing.

Probably should open new issues for anything found and close this issue. For posterity's sake, maybe also link any new issues here for tracking purposes?

**Comment by Xiphe** [comment]
> Please upgrade and begin stress testing.

Would like to help testing but am totally lost. Could anyone give an update on the current steps required to build for M1?

**Comment by lutzroeder** [comment]
1. Update to latest preview `electron-builder@22.10.4`
2. Run `npx electron-builder --mac --arm64` or `npx electron-builder --mac --universal`

There are some open PRs to fix follow-up issues.

**Comment by csett86** [comment]
@Xiphe If you are looking for the package.json updates required for universal builds, you could try this: https://github.com/csett86/jitsi-meet-electron/blob/universal/package.json

But as @lutzroeder said, there are some follup fixes currently required, see https://github.com/csett86/jitsi-meet-electron/tree/universal/patches

**Comment by Xiphe** [comment]
Thanks a lot! I decided to go with separate builds for mac in order to have less size overhead. 
Up until now everything worked well with `electron-builder@22.10.4` and running `npx electron-builder --mac --arm64 --x64` last think I need to test is if the auto update works correctly so that arm versions stay on their arch.

**Comment by transparentech** [comment]
Thanks a lot for everyone's work on this!

**Comment by transparentech** [comment]
But has anyone gotten the auto updating to work with arm64? I am able to build 2 separate versions for mac (x64 and arm64). But when I auto update the arm64 version, it updates to the x64 version. Is there something we need to do to tell the auto-updater which arch we need?

**Comment by davej** [comment]
@transparentech You will need to update electron-updater to the 4.3.7 beta. Works for me.

**Comment by mikeyyyzhao** [comment]
Anyone running into "ReferenceError: u is not defined" for auto update for arm64?

I have two separate s3 bucket, one for arm64 and one for x64. When I check for updates on arm64, the autoUpdater gets "ReferenceError: u is not defined" but not for x64. The only difference is how I build the app:

Arm64: `electron-builder --mac --arm64`
x64: `electron-builder --mac --x64`

Really appreciate all the hard work on this!

**Comment by AleksMeshkov** [comment]
Hm, strange... 
I have:

```
"electron": "^11.1.1",
"electron-builder": "^22.10.4",
"electron-updater": "^4.3.7",
```

and get this:

```
npx electron-builder --mac --universal
  • electron-builder  version=22.10.4 os=19.6.0
  • loaded configuration  file=package.json ("build" field)
  • writing effective config  file=release/builder-effective-config.yaml
  • rebuilding native dependencies  dependencies=diskusage@1.1.3, keytar@7.3.0 platform=darwin arch=x64
  • install prebuilt binary  name=keytar version=7.3.0 platform=darwin arch=x64
  • packaging       platform=darwin arch=x64 electron=11.1.1 appOutDir=release/mac-universal-x64
  • rebuilding native dependencies  dependencies=diskusage@1.1.3, keytar@7.3.0 platform=darwin arch=arm64
  • install prebuilt binary  name=keytar version=7.3.0 platform=darwin arch=arm64
  • build native dependency from sources  name=keytar
                                          version=7.3.0
                                          platform=darwin
                                          arch=arm64
                                          reason=prebuild-install failed with error (run with env DEBUG=electron-builder to get more information)
                                          error=prebuild-install info begin Prebuild-install version 6.0.0
    prebuild-install WARN install prebuilt binaries enforced with --force!
    prebuild-install WARN install prebuilt binaries may be out of date!
    prebuild-install info looking for cached prebuild @ /Users/alexey/.npm/_prebuilds/77ee54-keytar-v7.3.0-electron-v85-darwin-arm64.tar.gz
    prebuild-install http request GET https://github.com/atom/node-keytar/releases/download/v7.3.0/keytar-v7.3.0-electron-v85-darwin-arm64.tar.gz
    prebuild-install http 404 https://github.com/atom/node-keytar/releases/download/v7.3.0/keytar-v7.3.0-electron-v85-darwin-arm64.tar.gz
    prebuild-install WARN install No prebuilt binaries found (target=11.1.1 runtime=electron arch=arm64 libc= platform=darwin)
    
  • packaging       platform=darwin arch=arm64 electron=11.1.1 appOutDir=release/mac-universal-arm64
  • packaging       platform=darwin arch=universal electron=11.1.1 appOutDir=release/mac-universal
  ⨯ Expected all non-binary files to have identical SHAs when creating a universal build but "Contents/Info.plist" did not  stackTrace=
  Error: Expected all non-binary files to have identical SHAs when creating a universal build but "Contents/Info.plist" did not
      at exports.makeUniversalApp (/Users/alexey/Devel/slate/ssd-im-new/node_modules/@electron/universal/src/index.ts:107:15)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
      at MacPackager.doPack (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/macPackager.ts:116:9)
      at MacPackager.pack (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/macPackager.ts:167:7)
      at Packager.doBuild (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/packager.ts:434:9)
      at executeFinally (/Users/alexey/Devel/slate/ssd-im-new/node_modules/builder-util/src/promise.ts:12:14)
      at Packager._build (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/packager.ts:369:31)
      at Packager.build (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/packager.ts:333:12)
      at executeFinally (/Users/alexey/Devel/slate/ssd-im-new/node_modules/builder-util/src/promise.ts:12:14)

```

removing node_modules completely, then running npm install didn't help...

**Comment by quanglam2807** [comment]
> Hm, strange...
> I have:
> 
> ```
> "electron": "^11.1.1",
> "electron-builder": "^22.10.4",
> "electron-updater": "^4.3.7",
> ```
> 
> and when I run `npx electron-builder --mac --universal `, I get this:
> 
> ```
> npx electron-builder --mac --universal
>   • electron-builder  version=22.10.4 os=19.6.0
>   • loaded configuration  file=package.json ("build" field)
>   • writing effective config  file=release/builder-effective-config.yaml
>   • rebuilding native dependencies  dependencies=diskusage@1.1.3, keytar@7.3.0 platform=darwin arch=x64
>   • install prebuilt binary  name=keytar version=7.3.0 platform=darwin arch=x64
>   • packaging       platform=darwin arch=x64 electron=11.1.1 appOutDir=release/mac-universal-x64
>   • rebuilding native dependencies  dependencies=diskusage@1.1.3, keytar@7.3.0 platform=darwin arch=arm64
>   • install prebuilt binary  name=keytar version=7.3.0 platform=darwin arch=arm64
>   • build native dependency from sources  name=keytar
>                                           version=7.3.0
>                                           platform=darwin
>                                           arch=arm64
>                                           reason=prebuild-install failed with error (run with env DEBUG=electron-builder to get more information)
>                                           error=prebuild-install info begin Prebuild-install version 6.0.0
>     prebuild-install WARN install prebuilt binaries enforced with --force!
>     prebuild-install WARN install prebuilt binaries may be out of date!
>     prebuild-install info looking for cached prebuild @ /Users/alexey/.npm/_prebuilds/77ee54-keytar-v7.3.0-electron-v85-darwin-arm64.tar.gz
>     prebuild-install http request GET https://github.com/atom/node-keytar/releases/download/v7.3.0/keytar-v7.3.0-electron-v85-darwin-arm64.tar.gz
>     prebuild-install http 404 https://github.com/atom/node-keytar/releases/download/v7.3.0/keytar-v7.3.0-electron-v85-darwin-arm64.tar.gz
>     prebuild-install WARN install No prebuilt binaries found (target=11.1.1 runtime=electron arch=arm64 libc= platform=darwin)
>     
>   • packaging       platform=darwin arch=arm64 electron=11.1.1 appOutDir=release/mac-universal-arm64
>   • packaging       platform=darwin arch=universal electron=11.1.1 appOutDir=release/mac-universal
>  
> 
>  ⨯ Expected all non-binary files to have identical SHAs when creating a universal build but "Contents/Info.plist" did not  stackTrace=
>   Error: Expected all non-binary files to have identical SHAs when creating a universal build but "Contents/Info.plist" did not
>       at exports.makeUniversalApp (/Users/alexey/Devel/slate/ssd-im-new/node_modules/@electron/universal/src/index.ts:107:15)
>       at runMicrotasks (<anonymous>)
>       at processTicksAndRejections (internal/process/task_queues.js:93:5)
>       at MacPackager.doPack (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/macPackager.ts:116:9)
>       at MacPackager.pack (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/macPackager.ts:167:7)
>       at Packager.doBuild (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/packager.ts:434:9)
>       at executeFinally (/Users/alexey/Devel/slate/ssd-im-new/node_modules/builder-util/src/promise.ts:12:14)
>       at Packager._build (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/packager.ts:369:31)
>       at Packager.build (/Users/alexey/Devel/slate/ssd-im-new/node_modules/app-builder-lib/src/packager.ts:333:12)
>       at executeFinally (/Users/alexey/Devel/slate/ssd-im-new/node_modules/builder-util/src/promise.ts:12:14)
> ```
> 
> removing node_modules completely, then running npm install didn't help...

@lutzroeder This is similar to what I mentioned earlier (https://github.com/electron-userland/electron-builder/pull/5481#issuecomment-749875793)

@AleksMeshkov You can apply this patch to fix the bug: https://github.com/webcatalog/translatium/blob/master/patches/app-builder-lib%2B22.10.4.patch

**Comment by prabhatsaini91** [comment]
Is notarization working when built using `npx electron-builder --mac --arm64`?
I am unable to get my app notarized when building for arm64. However, x64 builds are notarized.

Using [**electron-notarize**](https://github.com/electron/electron-notarize) to perform notarization

**Comment by mmaietta** [comment]
I'd imagine that would be in an `afterSign` hook. Not sure if it's an electron-builder issue here. Here's my setup that validates the notarization as well:
```
async function notarizeMac(context: builder.AfterPackContext) {
    const appPath = path.join(context.appOutDir, `${ context.packager.appInfo.productFilename }.app`);
    await notarize({ appBundleId, appPath, ascProvider, appleId, appleIdPassword})
        .catch(err => { throw err; });
    executeCommand("spctl", ["-a", "-t", "open", "--context", "context:primary-signature", "-v", appPath])
}
```

**Comment by pliablepixels** [comment]
Hi, I am a little confused on what I am doing wrong.

This is my package.json

```
 "scripts": {
    "electron": "electron .",
    "dist-win": "./electron_js/sync_versions.sh && ./node_modules/.bin/electron-builder -w",
    "dist-lin": "./electron_js/sync_versions.sh && ./node_modules/.bin/electron-builder -l",
    "dist-mac": "./electron_js/sync_versions.sh && ./node_modules/.bin/electron-builder -m",
    "dist-all": "./electron_js/sync_versions.sh && ./node_modules/.bin/electron-builder -wlm"
  },
<snip>
 "build": {
    "artifactName": "${productName}-${version}-${os}-${arch}.${ext}",
    "appId": "com.pliablepixels.zmninja-pro",
    "productName": "zmninja-pro",
    "compression": "normal",
    "asar": true,
    "afterSign": "electron_js/notarize.js",
    "directories": {
      "buildResources": "electron_js/icons"
    },
<snip>
    "dmg": {
      "sign": false
    },
    "mac": {
      "category": "productivity",
      "target": {
        "target": "dmg",
        "arch": [
          "arm64",
          "x64"
        ]
      },
```

As you see, I have both arm64 and x67

The M1 mac app is built out, but only one dmg file is produced (x64) when I run `npm run dist-mac`
